planner/cascades/transformation_rules.go

```diff
@@ -1279,6 +1280,63 @@ func (r *PushTopNDownTiKVSingleGather) OnTransform(old *memo.ExprIter) (newExprs
 	return []*memo.GroupExpr{finalTopNExpr}, true, false, nil
 }
 
+// MergeAdjacentTopN merge adjacent TopN.
+type MergeAdjacentTopN struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentTopN creates a new Transformation MergeAdjacentTopN.
+// The pattern of this rule is `TopN->TopN->X`.
+func NewRuleMergeAdjacentTopN() Transformation {
+	rule := &MergeAdjacentTopN{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandTopN, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentTopN) Match(expr *memo.ExprIter) bool {
+	topN := expr.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	child := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalTopN)
+
+	// We can use this rule when the sort columns of parent TopN is a prefix of child TopN.
+	for i := 0; i < len(topN.ByItems); i++ {
+		if !topN.ByItems[i].Equal(topN.SCtx(), child.ByItems[i]) {
```

Comment:
Can we assert the `child.ByItems[i]` always exists?

Reply:
Good catch! And add some integration tests. PTAL @lzmhhh123